src/entities/map/ui/lib/cluster.ts

```diff
@@ -0,0 +1,207 @@
+        return {
+          lat: mean.lat + lat / this.markers.length,
+          lng: mean.lng + lng / this.markers.length,
+        };
+      },
+      { lat: 0, lng: 0 },
+    );
+
+    this.var = scaledLatLngList.reduce(
+      (variance, { lat, lng }) => {
+        return {
+          lat:
+            variance.lat +
+            Math.pow(lat - this.mean.lat, 2) / this.markers.length,
+          lng:
+            variance.lng +
+            Math.pow(lng - this.mean.lng, 2) / this.markers.length,
+        };
+      },
+      { lat: 0, lng: 0 },
+    );
+
+    this.std = {
+      lat: Math.sqrt(this.var.lat),
+      lng: Math.sqrt(this.var.lng),
+    };
+  }
+  private getZScore({ lat, lng }: LatLng) {
+    const scaledLatLng = this.minMaxScaling({ lat, lng });
+    return {
+      lat: Math.abs(scaledLatLng.lat - this.mean.lat) / this.std.lat,
+      lng: Math.abs(scaledLatLng.lng - this.mean.lng) / this.std.lng,
+    };
+  }
+  /**
+   * 특정 마커를 마커 버퍼에 추가 합니다.
+   * 이 때 추가된 마커가 이상값인지 판단하여 이상값 버퍼에 추가 합니다.
+   * ! 이상값으로 판단 된 마커는 클러스터 기준으로 이상값일 뿐
+   * ! 전체 데이터 기준으로 이상값이라고 판단 할 수 없습니다.
+   * ! 오히려 전체 데이터 기준으로 봤을 땐 해당 군집에 어울리는 값일 수 있습니다.
+   * ! 이상값을 판단하는 기준은 Z-Score가 3 이상인 경우로 설정 하였습니다. (3시그마 규칙)
+   */
+  addMarker(marker: T) {
+    const { lat, lng } = marker;
+    // 초기 시행 시에는 따로 Z-Score를 계산하지 않습니다.
+    if (this.std.lat === 0 || this.std.lng === 0) {
+      this.markers.push(marker);
+      return;
+    }
+    const zScore = this.getZScore({ lat, lng });
+    /**
+     * 새로운 군집을 형성 하기 위해 버퍼에 추가합니다.
+     * 이 때 새로운 군집을 형성하는 기준은 이상값이 아닌 마커들로 구성된 군집입니다.
+     */
+    if (zScore.lat > 3 || zScore.lng > 3) {
+      this.outliers.push(marker);
+      return;
+    }
+    this.markers.push(marker);
+  }
+  /**
+   * 마커의 중심점을 재조정 합니다.
+   * 이 때 마커의 중심점이 변경 되었다면 true를 반환 합니다.
+   */
+  revalidateCluster() {
+    // 재조정 전 클러스터의 중심점을 캐싱 합니다.
+    const prevCenter = { ...this.center };
+
+    // 클러스터의 마커들을 이용해 새로운 중심점을 계산합니다.
+    this.center = this.markers.reduce(
+      (center, { lat, lng }) => ({
+        lat: center.lat + lat / this.markers.length,
+        lng: center.lng + lng / this.markers.length,
+      }),
+      { lat: 0, lng: 0 },
+    );
+    // 새로운 중심점을 이용해 통계값을 계산합니다.
+    this.updateStatisticValue();
+    return (
+      prevCenter.lat !== this.center.lat || prevCenter.lng !== this.center.lng
+    );
+  }
+}
+
+export const useKMeansClustering = <T extends LatLng>(
+  NumOfCluster: number,
+  markers?: T[],
+) => {
+  const getCurrentBounds = useGetMapCurrentBounds();
+  const bounds = getCurrentBounds();
+  if (!markers) {
+    return [];
+  }
+
+  // K개의 클러스터를 생성합니다.
+  // TODO 휴리스틱한 방식으로 초기값 뽑기
+  const randomIndexMap: Record<number, boolean> = {};
+  const clusters = Array.from({ length: NumOfCluster }, () => {
+    let randomIndex;
+    do {
+      randomIndex = Math.floor(Math.random() * markers.length);
+    } while (randomIndexMap[randomIndex]);
+    {
+      randomIndexMap[randomIndex] = true;
+    }
```

Comment:
>feat : 랜덤한 인덱스를 뽑을 때 이전에 선택한 적 없던 인덱스를 뽑도록 do while 문을 추가해줬습니다.

여기서 이전에 선택한 적 없던 인덱스를 뽑히는 과정을 더 설명해주실 수 있나요?

Reply:
`do while` 문은 `do` 에 해당하는 블록을 우선 시행한 후 `while` 문을 돕니다 (이후엔 두 블록에 있는 블록을 와일문이 종료 될 때 까지 시행됩니다)

랜덤 인덱스 맵에는 이전 클러스터가 뽑은 인덱스 값이 키 값으로 존재하기에 내가 현재 뽑은 인덱스거 이전에 뽑은 인덱스라면 이전에 뽑은 인덱스가 아닐 때 까지 반복문을 돌며 인덱스를 뽑습니다 

와일문을 탈출했다면 그 인덱스는 랜덤 인덱스 맵에 존재하지 않는 값임이 보장 됩니다

Reply:
블록 하나를 없애야 할 것 같아요 👀 

```suggestion
    do {
      randomIndex = Math.floor(Math.random() * markers.length);
    } while (randomIndexMap[randomIndex]);
    
    randomIndexMap[randomIndex] = true;
    
```

Reply:
허거걱  그러네요 반영했습니다